doc/sphinx/proof-engine/vernacular-commands.rst

```diff
@@ -1308,7 +1181,55 @@ Inlining hints for the fast reduction machines
 Registering primitive operations
 ````````````````````````````````
 
-.. cmd:: Primitive @ident__1 := #@ident__2.
+.. cmd:: Primitive @ident {? : @term } := @register_token
```

Comment:
I'm really surprised that the list of registrable tokens is in the parser, but it is probably too much for the user documentation, especially as this is an internal command. So please add some rules to `common.edit_mlg` to define `register_token` as `#@ident` then revert this change (in particular, keep the indices).

Reply:
> I'm really surprised that the list of registrable tokens is in the parser

Agreed.

But we should say something about what values are valid, because `Primitive foo :=#bar` gives the error `Error: Syntax Error: Lexer: Undefined token`.

These don't look too great, either.  `#float64_type` and `#int63_type` are supposed to be valid:

```
Coq < Primitive foo := #float64_type.
Toplevel input, characters 17-18:
> Primitive foo := #float64_type.
>                  ^
Error: Syntax Error: Lexer: Undefined token

Coq < Primitive foo2 := #int63_type.
Toplevel input, characters 0-30:
> Primitive foo2 := #int63_type.
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error:
Anomaly
"File "kernel/primred.ml", line 15, characters 19-25: Assertion failed."
Please report at http://coq.inria.fr/bugs/.

Coq < Primitive foo := #int63_head0.
Toplevel input, characters 0-30:
> Primitive foo := #int63_head0.
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error:
The type int must be registered before this construction can be typechecked.
```

Reply:
The first two worked for me in 8.11.0 (no syntax error, no anomaly). I think you are testing this in a version of Coq that's too old (maybe Coq 8.10). The last one gave the error message, but this error message is reasonably clear to me. The fix is to add `Primitive bar := #int63_type.` before. Since these are internal commands to be used only by implementors of Coq's standard library, or standard library replacements, do you think it would be reasonable to refer to the theory files which contain these `Primitive` operations for the reference of which values are valid? The files are `theories/Floats/PrimFloat.v` and `theories/Numbers/Cyclic/Int63/Int63.v`. They contain 44 calls to the `Primitive` command, corresponding to the 44 accepted values.

See also the descriptions here: https://coq.github.io/doc/master/refman/language/gallina-extensions.html#primitive-floats

Reply:
Ugh, I installed 8.11 but ran coqtop in the 8.10 directory :-(.  Will update.

Reply:
Check it out.  Is the part about Register useful?  There is no Register in the referenced .v files.

```
   When running this command, the type of the primitive should be already known by
   the kernel (this is achieved through this command for primitive types and
   through the :cmd:`Register` command with the :g:`kernel` name-space for other
   types).
```

Reply:
I don't understand what you wrote here. The two files I referenced were about `Primitive` declarations only. As for the difference between `Primitive` and `Register`, it is pretty simple: `Register` is so that OCaml code knows about (and can manipulate) objects that were defined in a `.v` file whereas `Primitive` is to give names in a `.v` files to objects that are primitive because they were defined in OCaml code. There are many tactics that need to manipulate some specific Coq objects, and thus `git grep "Register" -- theories/` will give you a very long list.